util/dvsim/SimCfg.py

```diff
@@ -97,6 +97,8 @@ def __init__(self, flow_cfg_file, hjson_data, args, mk_config):
             self.en_build_modes.append("gui")
         if args.waves is not None:
             self.en_build_modes.append("waves")
+        else:
+            self.en_build_modes.append("waves_off")
```

Comment:
Instead of doing this, please could we add an explicit `--no-waves` argument to the toplevel or similar? The reason I ask is that it's really handy to be able to run 20 seeds of a test and then re-run just the ones that failed with `-w`. This change would force a rebuild in that case.

If I've understood correctly, this functionality is mostly to speed up our nightly tests, where there's no problem with adding weird extra options to the command line.

Reply:
> Instead of doing this, please could we add an explicit --no-waves argument to the toplevel or similar? The reason I ask is that it's really handy to be able to run 20 seeds of a test and then re-run just the ones that failed with -w. This change would force a rebuild in that case.

Looks like your usage is to build without waves, but then may run with `-w`. But in this case, your waveform loses some debug capability, for example, the fsdb won't contain the design hierarchy.
these options (-assert dbgopt+novpi) will not dump assertion failures in the FSDB (although assertion failure messages would come out during simulation). 

IIUC, you want to have these 3 usage.
1. no wave capability at all for build and sim. After the build, if we enable waves for sim, waves lose some debug info, like design hierarchy, and assertion dumping if PR is in.
2. enable wave capability for build and sim.
3. enable wave capability for build, then you may sim with or without waves. I think sim without waves in this case will be slower than case1.

Do you think we should have case3?

Reply:
I will prefer not adding `--no-waves` switch to dvsim. It makes the usage even more confusing (what would it then mean to not pass either?). 

I had an idea a long time ago in mind to support the usecase you present, @rswarbrick - in development mode, launch 2 builds in parallel - one with `debug` (enable waves, more observability into the design, etc) and one without. Tests would run is optimized mode using the one without, and dvsim will automatically rerun it if it failed, but this time it would use the build with `debug` enabled. This is the very first pending feature tracked for DV in #1727. That way, users don't need to manually rerun failures with waves. 

I tbh am not liking this change either. What would happen if we removed the `-debug_access+f` from non-waveform run? There is already a TODO on line 51 of `vcs.hjson`. Would that obviate the need for this optimization flag in non-wave run?

Reply:
I feel adding `-assert dbgopt+novpi` or not may not affect your debug. Since you don't use Verdi, I really don't know how you waves look like.
Can you see and use the SVA dump as follows? It shows where the violation is and the SVA can be expanded to see the related signals (you can certainly find these signals in RTL.)
![image](https://user-images.githubusercontent.com/49293026/156440088-c8278e9e-d100-47f7-8d6c-fae54146c65c.png)

Reply:
This is exactly the problem with the wording "waves-off" which implies the wave dump is completely turned off, which is actually not the case :) See my suggestion below.

Reply:
I feel we're not supposed to turn on waves for sim after building it without -w. We will lose some debug info if we use Verdi. Since Rupert doesn't use Verdi, it may make no difference to him.

So this mode is actually `build_with_partial_dump_capability` if we build without -w for VCS.

Reply:
Fair enough, but if I understand correctly this proposed change goes from "allow Rupert to dump the signals he cares about without rebuilding" to "force a rebuild". I'm not keen :-)